### PR TITLE
added select1 test with PDO update

### PR DESCRIPTION
**Pull request by smtakeda**
Added select1.phpt test case.

Updated PDO code in order to pass the test. It is not complete but works for now.
I'll revisit them as we add more tests.

### CODE REVIEW

**Comment by smtakeda** [comment]
@notkriswagner Fixed Snowflake client example codes so that they can accept HOST, PORT and PROTOCOL if specified.

File: config.m4 [hunk 1]
```diff
@@ -41,6 +41,8 @@ if test "$PHP_PDO_SNOWFLAKE" != "no"; then
   LDFLAGS="$LDFLAGS $SNOWFLAKE_CLIENT_DIR/deps-build/linux/curl/lib/libcurl.a"
   LDFLAGS="$LDFLAGS -Wl,--no-whole-archive"
 
+  CFLAGS="-std=c99 $CFLAGS"
```
Comment (smtakeda) [inline config.m4]: For PDO only at the moment. Some debug macro requires C99 features. Will drop this if this doesn't work on other platform.

File: pdo_snowflake.c [hunk 2]
```diff
@@ -27,11 +27,26 @@ ZEND_GET_MODULE(pdo_snowflake)
...
+  va_start (args, fmt);
+
+  FILE* fp = fopen("/tmp/php.log", "a");
+  fprintf(fp, "[%-6s] %-70s:%5d - ", severity, filename, line);
+  vfprintf (fp, fmt, args);
+  fprintf(fp, "\n");
+  fclose(fp);
+
+  va_end (args);
+}
```
Comment (smtakeda) [inline pdo_snowflake.c]: This is a temporary logger. Will integrate with Snowflake client later.

File: pdo_snowflake.c [hunk 3]
```diff
@@ -82,6 +97,9 @@ static PHP_GINIT_FUNCTION(pdo_snowflake)
 ZEND_TSRMLS_CACHE_UPDATE();
 #endif
 #endif
+#if PDO_DBG_ENABLED
+	pdo_snowflake_globals->debug = NULL;	/* The actual string */
```
Comment (smtakeda) [inline pdo_snowflake.c]: Trying to integrate with PHP ini config. Not tested yet.

File: php_pdo_snowflake_int.h [hunk 4]
```diff
@@ -27,7 +48,13 @@ typedef struct {
...
 typedef struct {
 	pdo_snowflake_db_handle *H;
-	SNOWFLAKE_STMT *stmt;
+	SNOWFLAKE_STMT          *stmt;
+	int64                   num_params;
+	SNOWFLAKE_BIND_INPUT    *params;
+	SNOWFLAKE_BIND_OUTPUT   *bound_result;
+	sf_bool					        *out_null;
+	zend_ulong				      *out_length;
+  void                   **data;
```
Comment (smtakeda) [inline php_pdo_snowflake_int.h]: Not finalized. Some could be pushed down to Snowflake client.

File: snowflake_driver.c [hunk 5]
```diff
@@ -38,21 +45,54 @@ static int snowflake_handle_closer(pdo_dbh_t *dbh) /* {{{ */
         pefree(H, dbh->is_persistent);
         dbh->driver_data = NULL;
     }
-    snowflake_global_term();
+    snowflake_global_term(); // TODO: should be one time per process
```
Comment (smtakeda) [inline snowflake_driver.c]: I believe these global init and term can move to pdo_snowflake.c such that they are called once per PDO init and term?

File: snowflake_driver.c [hunk 6]
```diff
@@ -38,21 +45,54 @@ static int snowflake_handle_closer(pdo_dbh_t *dbh) /* {{{ */
...
+ * you should use the pdo_parse_params() API to have PDO rewrite the query to
+ * take advantage of the support provided by your database.
+ *
+ * @param dbh Pointer to the database handle initialized by the handle factory
+ * @param sql Pointer to a character string containing the SQL statement to be prepared.
+ * @param sql_len The length of the SQL statement.
+ * @param stmt Pointer to the returned statement or NULL if an error occurs.
+ * @param driver_options Any driver specific/defined options.
+ * @return 1 if success or 0 if error occurs
+ */
```
Comment (smtakeda) [inline snowflake_driver.c]: Copied text from PHP doc.

File: snowflake_driver.c [hunk 7]
```diff
@@ -88,24 +128,10 @@ static int snowflake_handle_preparer(pdo_dbh_t *dbh, const char *sql, size_t sql
...
-//    if (S->num_params) {
-//        S->params_given = 0;
-//#if defined(PDO_USE_MYSQLND)
-//        S->params = NULL;
-//#else
-//        S->params = ecalloc(S->num_params, sizeof(MYSQL_BIND));
-//		S->in_null = ecalloc(S->num_params, sizeof(my_bool));
-//		S->in_length = ecalloc(S->num_params, sizeof(zend_ulong));
-//#endif
-//    }
```
Comment (smtakeda) [inline snowflake_driver.c]: Moved to exec

File: snowflake_stmt.c [hunk 8]
```diff
@@ -13,56 +13,322 @@
...
+        PDO_DBG_INF("prec: %d, scale: %d, nane: %s, type: %d",
+                    desc->precision, desc->scale, desc->name, desc->type);
+        if (desc->type == SF_TYPE_FIXED) {
+            if (desc->scale == 0) {
+                len = (size_t)desc->precision;
+            } else {
+                len = (size_t)desc->precision + desc->scale + 1;
+            }
+        } else if (desc->type == SF_TYPE_TEXT) {
+            len = 100;
```
Comment (smtakeda) [inline snowflake_stmt.c]: This will be updated when tested.

File: snowflake_stmt.c [hunk 9]
```diff
@@ -13,56 +13,322 @@
...
+        }
+        S->bound_result[i].max_length = len + 1; // change this
+        S->data[i] = ecalloc(len+1, sizeof(char)); // TODO: change this
```
Comment (smtakeda) [inline snowflake_stmt.c]: Actually the null termination is not required for PHP. So I want to remove +1 here.

File: snowflake_stmt.c [hunk 10]
```diff
@@ -13,56 +13,322 @@
...
+    /* error invalid column */
+    PDO_DBG_INF("ERROR 2");
+    PDO_DBG_RETURN(0);
+  }
+  for (i = 0; i < stmt->column_count; i++)
+  {
+    cols[i].precision = (zend_ulong)S->stmt->desc[i]->precision;
+    cols[i].maxlen = (size_t)S->stmt->desc[i]->byte_size;
+    cols[i].name = zend_string_init(S->stmt->desc[i]->name, strlen(S->stmt->desc[i]->name), 0);
+    cols[i].param_type = PDO_PARAM_STR;
```
Comment (smtakeda) [inline snowflake_stmt.c]: As the first step, I'm planning to use STR type for all cases.

File: config.m4 (hunk 1 above)
Comment (notkriswagner) [inline config.m4]: Is C99 required to get the PDO Driver to build? Otherwise I think this should be C89 or C90.

File: pdo_snowflake.c [hunk 11]
```diff
@@ -27,11 +27,26 @@ ZEND_GET_MODULE(pdo_snowflake)
...
 PHP_INI_BEGIN()
 #if PDO_DBG_ENABLED
-	STD_PHP_INI_ENTRY("pdo_snowflake.debug",	NULL, PHP_INI_SYSTEM, debug)
+		STD_PHP_INI_ENTRY("pdo_snowflake.debug",	NULL, PHP_INI_SYSTEM, OnUpdateString, debug, zend_pdo_snowflake_globals, pdo_snowflake_globals)
 #endif
 PHP_INI_END()
 /* }}} */
 
+/* TODO: adhoc logger until Snowflake client provides it. */
+void pdo_snowflake_log(int line, const char* filename, const char* severity, char *fmt, ...)
```
Comment (notkriswagner) [inline pdo_snowflake.c]: I'll add the logging file that I'm using to the include directory tomorrow

File: config.m4 (hunk 1 above)
Comment (smtakeda) [reply config.m4]: Yes, variable parameter macro requires C99. Actually mysql did. Will remove this if PDO won't be built with other platform.

Comment (notkriswagner) [reply config.m4]: Ah, interesting

**Pull request merged**